functional-tests/utils.py

```diff
@@ -432,3 +415,55 @@ def get_bridge_pubkey_from_cfg(cfg_params) -> str:
     op_x_only_pks = [convert_to_xonly_pk(pk) for pk in op_pks]
     agg_pubkey = musig_aggregate_pks(op_x_only_pks)
     return agg_pubkey
+
+
+def setup_root_logger() -> int:
+    """
+    reads `LOG_LEVEL` from the environment. Defaults to `NOTSET` if not provided.
+    """
+    log_level = os.getenv("LOG_LEVEL", "NOTSET").upper()
+    log_level = getattr(logging, log_level, logging.NOTSET)
+    # Configure the root logger
+    root_logger = logging.getLogger()
+    root_logger.setLevel(log_level)
+    return log_level
+
+
+def setup_test_loggers(datadir_root: str, tests: list[str], log_level: int):
+    """
+    Set up loggers for a list of test names, with log files in a logs directory.
+    - Configures both file and stream handlers for each test logger.
+    - Logs are stored in `<datadir_root>/logs/<test_name>.log`.
+
+    Parameters:
+        datadir_root (str): Root directory for logs.
+        log level (int): Log level
+        tests (list of str): List of test names to create loggers for.
+    """
+    # Create the logs directory
+    log_dir = os.path.join(datadir_root, "logs")
+    os.makedirs(log_dir, exist_ok=True)
+
+    # Common formatter
+    formatter = logging.Formatter(
+        "%(asctime)s - %(levelname)s - %(filename)s:%(lineno)d - %(message)s"
+    )
+    # Set up individual loggers for each test
+    for test_name in tests:
+        logger = logging.getLogger(test_name)
```

Comment:
This is essentially doing mutable global state, which you usually want to avoid.

More of this logic should be integrated into the runtime, which we can do now, see toplevel review comment.